Mathlib/Logic/Pairwise.lean

```diff
@@ -0,0 +1,88 @@
+/-
+Copyright (c) 2017 Johannes Hölzl. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johannes Hölzl
+-/
+import Mathlib.Logic.Function.Basic
+import Mathlib.Logic.Relation
+import Mathlib.Init.Set
+
+/-!
+# Relations holding pairwise
+
+This file defines pairwise relations.
+
+## Main declarations
+
+* `Pairwise`: `Pairwise r` states that `r i j` for all `i ≠ j`.
+* `Set.pairwise`: `s.pairwise r` states that `r i j` for all `i ≠ j` with `i, j ∈ s`.
+-/
+
+
+open Set Function
+
+variable {α β γ ι ι' : Type _} {r p q : α → α → Prop}
+
+section Pairwise
+
+variable {f g : ι → α} {s t u : Set α} {a b : α}
+
+/-- A relation `r` holds pairwise if `r i j` for all `i ≠ j`. -/
+def Pairwise (r : α → α → Prop) :=
+  ∀ ⦃i j⦄, i ≠ j → r i j
+#align pairwise Pairwise
+
+theorem Pairwise.mono (hr : Pairwise r) (h : ∀ ⦃i j⦄, r i j → p i j) : Pairwise p :=
+  fun _i _j hij => h <| hr hij
+#align pairwise.mono Pairwise.mono
+
+protected theorem Pairwise.eq (h : Pairwise r) : ¬r a b → a = b :=
+  not_imp_comm.1 <| @h _ _
+#align pairwise.eq Pairwise.eq
+
+theorem Function.injective_iff_pairwise_ne : Injective f ↔ Pairwise ((· ≠ ·) on f) :=
+  forall₂_congr fun _i _j => not_imp_not.symm
+#align function.injective_iff_pairwise_ne Function.injective_iff_pairwise_ne
+
+alias Function.injective_iff_pairwise_ne ↔ Function.Injective.pairwise_ne _
+
+namespace Set
+
+/-- The relation `r` holds pairwise on the set `s` if `r x y` for all *distinct* `x y ∈ s`. -/
+protected def Pairwise (s : Set α) (r : α → α → Prop) :=
+  ∀ ⦃x⦄, x ∈ s → ∀ ⦃y⦄, y ∈ s → x ≠ y → r x y
+#align set.pairwise Set.Pairwise
+
+theorem pairwise_of_forall (s : Set α) (r : α → α → Prop) (h : ∀ a b, r a b) : s.Pairwise r :=
+  fun a _ b _ _ => h a b
+#align set.pairwise_of_forall Set.pairwise_of_forall
+
+theorem Pairwise.imp_on (h : s.Pairwise r) (hrp : s.Pairwise fun ⦃a b : α⦄ => r a b → p a b) :
+    s.Pairwise p :=
+  fun _a ha _b hb hab => hrp ha hb hab <| h ha hb hab
+#align set.pairwise.imp_on Set.Pairwise.imp_on
+
+theorem Pairwise.imp (h : s.Pairwise r) (hpq : ∀ ⦃a b : α⦄, r a b → p a b) : s.Pairwise p :=
+  h.imp_on <| pairwise_of_forall s _ hpq
+#align set.pairwise.imp Set.Pairwise.imp
+
+protected theorem Pairwise.eq (hs : s.Pairwise r) (ha : a ∈ s) (hb : b ∈ s) (h : ¬r a b) : a = b :=
+  of_not_not fun hab => h <| hs ha hb hab
+#align set.pairwise.eq Set.Pairwise.eq
+
+theorem _root_.Reflexive.set_pairwise_iff (hr : Reflexive r) :
+    s.Pairwise r ↔ ∀ ⦃a⦄, a ∈ s → ∀ ⦃b⦄, b ∈ s → r a b :=
+  forall₄_congr fun a _ _ _ => or_iff_not_imp_left.symm.trans <| or_iff_right_of_imp <| Eq.ndrec <|
+    hr a
+#align set._root_.reflexive.set_pairwise_iff set._root_.Reflexive.set_pairwise_iff
```

Comment:
I don't think there should ever be `_root_` in an #align. They are fully qualified names.

Reply:
```suggestion
#align reflexive.set_pairwise_iff Reflexive.set_pairwise_iff
```